File: solutions/lua/dnd-character/1/dnd-character.lua
local Character = {}

math.randomseed(os.time())

local function ability(scores)
	table.sort(scores, function(a, b) return a > b end)
	local acc = 0
	for i = 1, 3 do
		acc = acc + scores[i]
	end
	return acc
end

local function roll_dice()
	local rolls = {}
	for _ = 1, 4 do
		table.insert(rolls, math.random(1, 6))
	end
	return rolls
end

local function modifier(input)
	local tmp = (input - 10)
	return (tmp - (tmp % 2)) / 2
end


MAGIC = 12

function Character:new(name)
	local char = {
		name = name,
		strength = MAGIC,
		dexterity = MAGIC,
		constitution = MAGIC,
		intelligence = MAGIC,
		wisdom = MAGIC,
		charisma = MAGIC,
		hitpoints = 10 + modifier(MAGIC),
	}
	return char
end

return { Character = Character, ability = ability, roll_dice = roll_dice, modifier = modifier }

File: solutions/ocaml/dnd-character/1/dnd_character.ml
open Base

type character = {
  charisma : int;
  constitution : int;
  dexterity : int;
  hitpoints : int;
  intelligence : int;
  strength : int;
  wisdom : int;
}

let ability () = 3 + Random.int 16

let modifier ~(score : int) : int =
  Float.(iround_down_exn (((of_int score) - 10.0) / 2.0))

let generate_character () : character =
  let constitution = ability () in
  { charisma = ability ();
    constitution = constitution;
    dexterity = ability ();
    hitpoints = 10 + modifier ~score:constitution;
    intelligence = ability ();
    strength = ability ();
    wisdom = ability ();
  }
